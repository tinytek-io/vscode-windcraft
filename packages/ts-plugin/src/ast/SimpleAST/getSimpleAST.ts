import type * as ts from "typescript/lib/tsserverlibrary";
import { log } from "../../lib/log";
import { SimpleAST } from "./SimpleAST";

export function getSimpleAST(
  fileName: string,
  program: ts.Program,
  typescript: typeof ts
) {
  const file = program.getSourceFile(fileName);
  if (!file) {
    return undefined;
  }

  const nodeWalker = (node: ts.Node, indent = ""): SimpleAST => {
    const iLog = (msg: string) => {}; //  log(`${indent}${msg}`);

    const result: SimpleAST = {
      kind: typescript.SyntaxKind[node.kind],
      position: {
        start: node.getStart(),
        end: node.getEnd(),
      },
    };

    iLog(`kind: ${result.kind}`);

    if (typescript.isFunctionDeclaration(node)) {
      result.name = node.name?.getText();
    }
    if (typescript.isJsxSelfClosingElement(node)) {
      result.name = node.tagName.getText();
    }
    if (
      typescript.isJsxOpeningElement(node) ||
      typescript.isJsxSelfClosingElement(node) ||
      typescript.isJsxClosingElement(node)
    ) {
      //  result.name = node.tagName.getText();
    }
    if (typescript.isJsxElement(node)) {
      result.name = node.openingElement.tagName.getText();
    }
    if (typescript.isJsxAttribute(node)) {
      result.name = node.name.getText();
    }
    if (typescript.isStringLiteral(node)) {
      result.name = node.text;
    }

    if (result.name) {
      iLog(result.name);
    }

    typescript.forEachChild(node, (n) => {
      if (!result.children) {
        result.children = [];
      }
      result.children.push(nodeWalker(n, indent + "  "));
    });

    return result;
  };

  return nodeWalker(file);
}
