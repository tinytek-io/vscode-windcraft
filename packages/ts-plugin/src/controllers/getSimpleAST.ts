import type { Request, Response } from "express";
import { log } from "../lib/log";

import { getSimpleAST } from "../ast/SimpleAST/getSimpleAST";
import { z } from "zod";

const getSimpleASTRequestDtoSchema = z.object({
  fileName: z.string(),
});

export type GetSimpleASTRequestDto = z.infer<
  typeof getSimpleASTRequestDtoSchema
>;

export function getSimpleASTController(req: Request, res: Response) {
  log("Received request '/' - getSimpleASTController");
  const { success, error, data } = getSimpleASTRequestDtoSchema.safeParse(
    req.body
  );

  if (!success) {
    log(`Sending response '/' - Invalid request`);
    res.status(400).send({ message: "Invalid request", error });
    return;
  }

  const { fileName } = data;

  log(`Processing request '/' - fileName: "${fileName}"`);

  const program = req.languageService.getProgram();
  if (!program) {
    log(`Sending response '/' - Program not found`);
    res.status(404).send({ message: "Program not found" });
    return;
  }

  const simpleAST = getSimpleAST(fileName, program, req.typescript);

  if (!simpleAST) {
    log(`Sending response '/' - File not found "${fileName}"`);
    res.status(404).send({ message: "File not found" });
  } else {
    log("Sending response '/'");
    res.send(simpleAST);
  }
}
