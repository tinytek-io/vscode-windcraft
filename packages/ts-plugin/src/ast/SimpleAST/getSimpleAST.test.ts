import { describe, it, expect } from "bun:test";
import * as ts from "typescript";
import { getSimpleAST } from "./getSimpleAST";
import { createTestProgram, testFile001 } from "../../test/createTestProgram";

describe("SimpleASTService", () => {
  const program = createTestProgram([testFile001.button]);

  describe("program", () => {
    it("program should contain nodes", () => {
      expect(program.getNodeCount()).toBeGreaterThan(0);
    });

    it("program should contain source files", () => {
      const sourceFiles = program.getSourceFiles().map((sf) => sf.fileName);
      expect(sourceFiles).toContain(testFile001.button);
    });
  });

  describe("getSimpleAST for test project 001", () => {
    Object.keys(testFile001).forEach((key) => {
      it(`should return SimpleAST for "${key}"`, () => {
        expect(getSimpleAST(testFile001.button, program, ts)).toMatchSnapshot();
      });
    });
  });
});
