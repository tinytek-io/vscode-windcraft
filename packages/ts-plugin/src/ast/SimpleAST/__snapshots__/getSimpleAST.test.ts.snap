// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleASTService getSimpleAST for test project 001 should return SimpleAST for "button" 1`] = `
{
  "children": [
    {
      "children": [
        {
          "kind": "ExportKeyword",
          "position": {
            "end": 6,
            "start": 0,
          },
        },
        {
          "kind": "Identifier",
          "position": {
            "end": 23,
            "start": 12,
          },
        },
        {
          "children": [
            {
              "children": [
                {
                  "kind": "Identifier",
                  "position": {
                    "end": 37,
                    "start": 30,
                  },
                },
                {
                  "children": [
                    {
                      "kind": "VoidKeyword",
                      "position": {
                        "end": 49,
                        "start": 45,
                      },
                    },
                  ],
                  "kind": "FunctionType",
                  "position": {
                    "end": 49,
                    "start": 39,
                  },
                },
              ],
              "kind": "PropertySignature",
              "position": {
                "end": 49,
                "start": 30,
              },
            },
            {
              "children": [
                {
                  "kind": "Identifier",
                  "position": {
                    "end": 60,
                    "start": 52,
                  },
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "kind": "Identifier",
                          "position": {
                            "end": 67,
                            "start": 62,
                          },
                        },
                        {
                          "kind": "Identifier",
                          "position": {
                            "end": 77,
                            "start": 68,
                          },
                        },
                      ],
                      "kind": "FirstNode",
                      "position": {
                        "end": 77,
                        "start": 62,
                      },
                    },
                  ],
                  "kind": "TypeReference",
                  "position": {
                    "end": 77,
                    "start": 62,
                  },
                },
              ],
              "kind": "PropertySignature",
              "position": {
                "end": 77,
                "start": 52,
              },
            },
          ],
          "kind": "TypeLiteral",
          "position": {
            "end": 79,
            "start": 26,
          },
        },
      ],
      "kind": "TypeAliasDeclaration",
      "position": {
        "end": 79,
        "start": 0,
      },
    },
    {
      "children": [
        {
          "kind": "ExportKeyword",
          "position": {
            "end": 87,
            "start": 81,
          },
        },
        {
          "kind": "Identifier",
          "position": {
            "end": 103,
            "start": 97,
          },
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "kind": "Identifier",
                      "position": {
                        "end": 113,
                        "start": 106,
                      },
                    },
                  ],
                  "kind": "BindingElement",
                  "position": {
                    "end": 113,
                    "start": 106,
                  },
                },
                {
                  "children": [
                    {
                      "kind": "Identifier",
                      "position": {
                        "end": 123,
                        "start": 115,
                      },
                    },
                  ],
                  "kind": "BindingElement",
                  "position": {
                    "end": 123,
                    "start": 115,
                  },
                },
              ],
              "kind": "ObjectBindingPattern",
              "position": {
                "end": 125,
                "start": 104,
              },
            },
            {
              "children": [
                {
                  "kind": "Identifier",
                  "position": {
                    "end": 138,
                    "start": 127,
                  },
                },
              ],
              "kind": "TypeReference",
              "position": {
                "end": 138,
                "start": 127,
              },
            },
          ],
          "kind": "Parameter",
          "position": {
            "end": 138,
            "start": 104,
          },
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "kind": "Identifier",
                              "position": {
                                "end": 164,
                                "start": 158,
                              },
                            },
                            {
                              "children": [
                                {
                                  "children": [
                                    {
                                      "kind": "Identifier",
                                      "position": {
                                        "end": 174,
                                        "start": 165,
                                      },
                                    },
                                    {
                                      "kind": "StringLiteral",
                                      "name": "flex",
                                      "position": {
                                        "end": 181,
                                        "start": 175,
                                      },
                                    },
                                  ],
                                  "kind": "JsxAttribute",
                                  "name": "className",
                                  "position": {
                                    "end": 181,
                                    "start": 165,
                                  },
                                },
                                {
                                  "children": [
                                    {
                                      "kind": "Identifier",
                                      "position": {
                                        "end": 189,
                                        "start": 182,
                                      },
                                    },
                                    {
                                      "children": [
                                        {
                                          "kind": "Identifier",
                                          "position": {
                                            "end": 198,
                                            "start": 191,
                                          },
                                        },
                                      ],
                                      "kind": "JsxExpression",
                                      "position": {
                                        "end": 199,
                                        "start": 190,
                                      },
                                    },
                                  ],
                                  "kind": "JsxAttribute",
                                  "name": "onClick",
                                  "position": {
                                    "end": 199,
                                    "start": 182,
                                  },
                                },
                              ],
                              "kind": "JsxAttributes",
                              "position": {
                                "end": 199,
                                "start": 165,
                              },
                            },
                          ],
                          "kind": "JsxOpeningElement",
                          "position": {
                            "end": 200,
                            "start": 157,
                          },
                        },
                        {
                          "kind": "JsxText",
                          "position": {
                            "end": 207,
                            "start": 207,
                          },
                        },
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "kind": "Identifier",
                                  "position": {
                                    "end": 211,
                                    "start": 208,
                                  },
                                },
                                {
                                  "children": [
                                    {
                                      "children": [
                                        {
                                          "kind": "Identifier",
                                          "position": {
                                            "end": 221,
                                            "start": 212,
                                          },
                                        },
                                        {
                                          "kind": "StringLiteral",
                                          "name": "flex flex-col",
                                          "position": {
                                            "end": 237,
                                            "start": 222,
                                          },
                                        },
                                      ],
                                      "kind": "JsxAttribute",
                                      "name": "className",
                                      "position": {
                                        "end": 237,
                                        "start": 212,
                                      },
                                    },
                                  ],
                                  "kind": "JsxAttributes",
                                  "position": {
                                    "end": 237,
                                    "start": 212,
                                  },
                                },
                              ],
                              "kind": "JsxOpeningElement",
                              "position": {
                                "end": 238,
                                "start": 207,
                              },
                            },
                            {
                              "kind": "JsxText",
                              "position": {
                                "end": 242,
                                "start": 238,
                              },
                            },
                            {
                              "children": [
                                {
                                  "kind": "Identifier",
                                  "position": {
                                    "end": 247,
                                    "start": 244,
                                  },
                                },
                              ],
                              "kind": "JsxClosingElement",
                              "position": {
                                "end": 248,
                                "start": 242,
                              },
                            },
                          ],
                          "kind": "JsxElement",
                          "name": "div",
                          "position": {
                            "end": 248,
                            "start": 207,
                          },
                        },
                        {
                          "kind": "JsxText",
                          "position": {
                            "end": 255,
                            "start": 255,
                          },
                        },
                        {
                          "children": [
                            {
                              "kind": "Identifier",
                              "position": {
                                "end": 259,
                                "start": 256,
                              },
                            },
                            {
                              "children": [
                                {
                                  "children": [
                                    {
                                      "kind": "Identifier",
                                      "position": {
                                        "end": 263,
                                        "start": 260,
                                      },
                                    },
                                    {
                                      "kind": "StringLiteral",
                                      "name": "test",
                                      "position": {
                                        "end": 270,
                                        "start": 264,
                                      },
                                    },
                                  ],
                                  "kind": "JsxAttribute",
                                  "name": "src",
                                  "position": {
                                    "end": 270,
                                    "start": 260,
                                  },
                                },
                                {
                                  "children": [
                                    {
                                      "kind": "Identifier",
                                      "position": {
                                        "end": 280,
                                        "start": 271,
                                      },
                                    },
                                    {
                                      "kind": "StringLiteral",
                                      "name": "w-full",
                                      "position": {
                                        "end": 289,
                                        "start": 281,
                                      },
                                    },
                                  ],
                                  "kind": "JsxAttribute",
                                  "name": "className",
                                  "position": {
                                    "end": 289,
                                    "start": 271,
                                  },
                                },
                              ],
                              "kind": "JsxAttributes",
                              "position": {
                                "end": 289,
                                "start": 260,
                              },
                            },
                          ],
                          "kind": "JsxSelfClosingElement",
                          "name": "img",
                          "position": {
                            "end": 292,
                            "start": 255,
                          },
                        },
                        {
                          "kind": "JsxText",
                          "position": {
                            "end": 299,
                            "start": 299,
                          },
                        },
                        {
                          "children": [
                            {
                              "kind": "Identifier",
                              "position": {
                                "end": 308,
                                "start": 300,
                              },
                            },
                          ],
                          "kind": "JsxExpression",
                          "position": {
                            "end": 309,
                            "start": 299,
                          },
                        },
                        {
                          "kind": "JsxText",
                          "position": {
                            "end": 314,
                            "start": 314,
                          },
                        },
                        {
                          "children": [
                            {
                              "kind": "Identifier",
                              "position": {
                                "end": 322,
                                "start": 316,
                              },
                            },
                          ],
                          "kind": "JsxClosingElement",
                          "position": {
                            "end": 323,
                            "start": 314,
                          },
                        },
                      ],
                      "kind": "JsxElement",
                      "name": "button",
                      "position": {
                        "end": 323,
                        "start": 157,
                      },
                    },
                  ],
                  "kind": "ParenthesizedExpression",
                  "position": {
                    "end": 327,
                    "start": 151,
                  },
                },
              ],
              "kind": "ReturnStatement",
              "position": {
                "end": 328,
                "start": 144,
              },
            },
          ],
          "kind": "Block",
          "position": {
            "end": 330,
            "start": 140,
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "name": "Button",
      "position": {
        "end": 330,
        "start": 81,
      },
    },
    {
      "kind": "EndOfFileToken",
      "position": {
        "end": 330,
        "start": 330,
      },
    },
  ],
  "kind": "SourceFile",
  "position": {
    "end": 330,
    "start": 0,
  },
}
`;
